File: jar/src/main/java/jline/api/mam/m3pp/M3pp22_interleave_fitc.java
/**
 * @file M3PP(2,2) interleaved superposition fitting
 *
 * Implements lumped superposition of multiple M3PP(2,2) processes using interleaved
 * parameter fitting.
 *
 * @since LINE 3.0
 */
package jline.api.mam.m3pp;

import java.util.ArrayList;
import java.util.List;

import jline.util.Pair;
import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class M3pp22_interleave_fitc {
    private M3pp22_interleave_fitc() {}

    /**
     * Fits L pairs of classes into a single MMAP.
     *
     * @return Pair of (lumped superposition MMAP, list of component M3PPs)
     */
    public static Pair<MatrixCell, List<MatrixCell>> m3pp22_interleave_fitc(
            double[][] av, double[] btv, double[] binfv, double[] stv, double t) {
        int L = av.length;
        if (btv.length != L) throw new IllegalArgumentException("btv must have length L");
        if (binfv.length != L) throw new IllegalArgumentException("binfv must have length L");
        if (stv.length != L) throw new IllegalArgumentException("stv must have length L");
        for (double[] row : av) {
            if (row.length != 2) throw new IllegalArgumentException("av must be L x 2 matrix");
        }

        double[] upperBounds = new double[L];
        double[] lowerBounds = new double[L];
        double[] dValues = new double[L];

        for (int i = 0; i < L; i++) {
            double totalRate = av[i][0] + av[i][1];
            double d = computeD(btv[i], binfv[i], t);
            double z = (binfv[i] - 1.0) * d * d * d * totalRate;
            double u = d * z / (2 * totalRate * totalRate * d * d + z);

            upperBounds[i] = u;
            lowerBounds[i] = d;
            dValues[i] = d;
        }

        double[][] offDiagonalRates = solveOffDiagonalOptimization(upperBounds, lowerBounds, dValues, L);

        List<MatrixCell> m3pps = new ArrayList<MatrixCell>();

        for (int i = 0; i < L; i++) {
            double r1 = 0.0;
            for (int it = 0; it < L; it++) {
                if (it >= i) r1 += offDiagonalRates[0][it];
            }
            double r2 = 0.0;
            for (int it = 0; it < L; it++) {
                if (it <= i) r2 += offDiagonalRates[1][it];
            }

            MatrixCell mmpp = new MatrixCell(2);
            mmpp.set(0, new Matrix(2, 2));
            mmpp.set(1, new Matrix(2, 2));

            mmpp.get(0).set(0, 1, r1);
            mmpp.get(0).set(1, 0, r2);

            double totalRate = av[i][0] + av[i][1];
            double d = r1 + r2;
            double z = (binfv[i] - 1.0) * d * d * d * totalRate;
            double delta = Math.sqrt(z / (2 * r1 * r2));
            double lambda2 = totalRate - r2 / d * delta;
            double lambda1 = lambda2 + delta;

            mmpp.get(1).set(0, 0, lambda1);
            mmpp.get(1).set(1, 1, lambda2);

            mmpp.get(0).set(0, 0, -(mmpp.get(0).get(0, 1) + mmpp.get(1).get(0, 0)));
            mmpp.get(0).set(1, 1, -(mmpp.get(0).get(1, 0) + mmpp.get(1).get(1, 1)));

            double variance = computeMmppVariance(mmpp, t);
            System.out.println("MMPP " + i + " - Var(t): " + variance);

            MatrixCell m3pp = m3pp22_fitc_approx_cov_multiclass(mmpp, av[i], stv[i], t);
            m3pps.add(m3pp);
        }

        MatrixCell lumped = M3pp2m_interleave.m3pp2m_interleave(m3pps);

        return new Pair<MatrixCell, List<MatrixCell>>(lumped, m3pps);
    }

    private static double computeD(double bt1, double binf, double t1) {
        if (!(binf > bt1 && bt1 > 1.0)) {
            throw new IllegalArgumentException(
                    "No solution, infeasible IDC(t): IDC(" + t1 + ") = " + bt1 + ", IDC(inf) = " + binf);
        }
        double c = (binf - 1.0) / (binf - bt1);
        double z = -c * Math.exp(-c);
        double w = lambertW(z, 100, 1e-12);
        return (w + c) / t1;
    }

    private static double lambertW(double z, int maxIter, double tolerance) {
        double w = (z > -0.1) ? z : -1.0;

        for (int iter = 0; iter < maxIter; iter++) {
            double ew = Math.exp(w);
            double wew = w * ew;
            double f = wew - z;
            double df = ew * (w + 1.0);

            if (Math.abs(df) < tolerance) break;

            double delta = f / df;
            w -= delta;

            if (Math.abs(delta) < tolerance) break;
        }
        return w;
    }

    private static double[][] solveOffDiagonalOptimization(double[] upperBounds,
                                                           double[] lowerBounds,
                                                           double[] dValues, int L) {
        double[][] r = new double[2][L];
        for (int i = 0; i < L; i++) {
            double totalD = dValues[i];
            double minU = Math.min(upperBounds[i], totalD * 0.6);
            double minL = Math.min(lowerBounds[i] - minU, totalD * 0.4);

            r[0][i] = minU;
            r[1][i] = Math.max(0.0, minL);

            double sum = 0.0;
            for (int it = 0; it < L; it++) {
                if (it >= i) sum += r[0][it];
                if (it <= i) sum += r[1][it];
            }

            if (Math.abs(sum - totalD) > 1e-6) {
                double adjustment = (totalD - sum) / 2.0;
                r[0][i] += adjustment;
                r[1][i] += adjustment;
            }
        }
        return r;
    }

    private static double computeMmppVariance(MatrixCell mmpp, double t) {
        Matrix D0 = mmpp.get(0);
        Matrix D1 = mmpp.get(1);

        double lambda1 = D1.get(0, 0);
        double lambda2 = D1.get(1, 1);
        double r12 = D0.get(0, 1);
        double r21 = D0.get(1, 0);

        double meanRate = (lambda1 * r21 + lambda2 * r12) / (r12 + r21);
        return meanRate * t * (1.0 + Math.pow(lambda1 - lambda2, 2) / Math.pow(r12 + r21, 2));
    }

    private static MatrixCell m3pp22_fitc_approx_cov_multiclass(MatrixCell mmpp,
                                                                double[] classRates,
                                                                double covariance, double t) {
        MatrixCell m3pp = new MatrixCell(4);
        m3pp.set(0, mmpp.get(0));
        m3pp.set(1, mmpp.get(1));

        double totalRate = classRates[0] + classRates[1];
        if (totalRate > 0) {
            double prop1 = classRates[0] / totalRate;
            double prop2 = classRates[1] / totalRate;

            m3pp.set(2, scaleMatrix(mmpp.get(1), prop1));
            m3pp.set(3, scaleMatrix(mmpp.get(1), prop2));
        } else {
            m3pp.set(2, new Matrix(2, 2));
            m3pp.set(3, new Matrix(2, 2));
        }
        return m3pp;
    }

    private static Matrix scaleMatrix(Matrix m, double factor) {
        Matrix scaled = new Matrix(m.getNumRows(), m.getNumCols());
        for (int i = 0; i < m.getNumRows(); i++) {
            for (int j = 0; j < m.getNumCols(); j++) {
                scaled.set(i, j, m.get(i, j) * factor);
            }
        }
        return scaled;
    }
}

File: jar/src/main/kotlin/jline/api/mam/m3pp/M3pp22_interleave_fitc.kt
/**
 * @file M3PP(2,2) interleaved superposition fitting
 * 
 * Implements lumped superposition of multiple M3PP(2,2) processes using interleaved
 * parameter fitting. Combines L different two-class arrival processes into a single
 * MMAP through optimization-based parameter estimation and Lambert W function solutions.
 * 
 * @since LINE 3.0
 */
package jline.api.mam.m3pp

import jline.util.matrix.MatrixCell
import jline.util.matrix.Matrix
import kotlin.math.*

/**
 * Fits L pairs of classes into a single MMAP, obtained by lumped superposition 
 * of L different M3PP[2] processes.
 * 
 * @param av Matrix of size L x 2 containing the per-class rates
 * @param btv Vector of length L containing the IDC at resolution t for each pair of classes
 * @param binfv Vector of length L containing the asymptotic IDC for each pair of classes
 * @param stv Vector of length L containing the count covariance at resolution t between each pair of classes
 * @param t Time resolution
 * @return Pair of (lumped superposition MMAP, list of component M3PPs)
 */
fun m3pp22_interleave_fitc(
    av: Array<DoubleArray>,
    btv: DoubleArray,
    binfv: DoubleArray,
    stv: DoubleArray,
    t: Double
): Pair<MatrixCell, List<MatrixCell>> {
    
    val L = av.size
    require(btv.size == L) { "btv must have length L" }
    require(binfv.size == L) { "binfv must have length L" }
    require(stv.size == L) { "stv must have length L" }
    require(av.all { it.size == 2 }) { "av must be L x 2 matrix" }
    
    // Compute bounds for upper off-diagonal elements of each MMPP(2)
    val upperBounds = DoubleArray(L)
    val lowerBounds = DoubleArray(L)
    val dValues = DoubleArray(L)
    
    for (i in 0 until L) {
        val totalRate = av[i].sum()
        val d = computeD(btv[i], binfv[i], t)
        val z = (binfv[i] - 1.0) * d * d * d * totalRate
        val u = d * z / (2 * totalRate * totalRate * d * d + z)
        
        upperBounds[i] = u
        lowerBounds[i] = d
        dValues[i] = d
    }
    
    // Solve linear program to find feasible off-diagonal elements
    val offDiagonalRates = solveOffDiagonalOptimization(upperBounds, lowerBounds, dValues, L)
    
    // Fit individual M3PP[2] processes
    val m3pps = mutableListOf<MatrixCell>()
    
    for (i in 0 until L) {
        // Compute transition rates for this MMPP
        val r1 = (0 until L).filter { it >= i }.sumOf { offDiagonalRates[0][it] }
        val r2 = (0 until L).filter { it <= i }.sumOf { offDiagonalRates[1][it] }
        
        // Create underlying MMPP(2)
        val mmpp = MatrixCell(2)
        mmpp[0] = Matrix(2, 2) // D0
        mmpp[1] = Matrix(2, 2) // D1
        
        mmpp[0][0, 1] = r1
        mmpp[0][1, 0] = r2
        
        val totalRate = av[i].sum()
        val d = r1 + r2
        val z = (binfv[i] - 1.0) * d * d * d * totalRate
        val delta = sqrt(z / (2 * r1 * r2))
        val lambda2 = totalRate - r2 / d * delta
        val lambda1 = lambda2 + delta
        
        mmpp[1][0, 0] = lambda1
        mmpp[1][1, 1] = lambda2
        
        // Set diagonal elements of D0
        mmpp[0][0, 0] = -(mmpp[0][0, 1] + mmpp[1][0, 0])
        mmpp[0][1, 1] = -(mmpp[0][1, 0] + mmpp[1][1, 1])
        
        // Verify variance (for debugging)
        val variance = computeMmppVariance(mmpp, t)
        println("MMPP $i - Var(t): $variance")
        
        // Fit M3PP(2,2) from this MMPP
        val m3pp = m3pp22_fitc_approx_cov_multiclass(mmpp, av[i], stv[i], t)
        m3pps.add(m3pp)
    }
    
    // Create lumped superposition using interleaving
    val lumped = m3pp2m_interleave(m3pps)
    
    return Pair(lumped, m3pps)
}

/**
 * Solve non-linear equation to compute d parameter for MMPP fitting
 */
private fun computeD(bt1: Double, binf: Double, t1: Double): Double {
    if (!(binf > bt1 && bt1 > 1.0)) {
        throw IllegalArgumentException(
            "No solution, infeasible IDC(t): IDC($t1) = $bt1, IDC(inf) = $binf"
        )
    }
    
    val c = (binf - 1.0) / (binf - bt1)
    
    // Solve using Newton's method for Lambert W function approximation
    // We need to solve: d = (W(-c*exp(-c)) + c) / t1
    val z = -c * exp(-c)
    val w = lambertW(z)
    val d = (w + c) / t1
    
    return d
}

/**
 * Approximate Lambert W function using Newton's method
 */
private fun lambertW(z: Double, maxIter: Int = 100, tolerance: Double = 1e-12): Double {
    var w = if (z > -0.1) z else -1.0 // Initial guess
    
    for (iter in 0 until maxIter) {
        val ew = exp(w)
        val wew = w * ew
        val f = wew - z
        val df = ew * (w + 1.0)
        
        if (abs(df) < tolerance) break
        
        val delta = f / df
        w -= delta
        
        if (abs(delta) < tolerance) break
    }
    
    return w
}

/**
 * Solve linear programming problem for off-diagonal rates
 * This is a simplified version - in practice would use proper LP solver
 */
private fun solveOffDiagonalOptimization(
    upperBounds: DoubleArray,
    lowerBounds: DoubleArray,
    dValues: DoubleArray,
    L: Int
): Array<DoubleArray> {
    
    // Simplified solution: distribute rates proportionally
    // In practice, would use proper linear programming solver
    
    val r = Array(2) { DoubleArray(L) }
    
    // Simple feasible solution
    for (i in 0 until L) {
        val totalD = dValues[i]
        val minU = minOf(upperBounds[i], totalD * 0.6)
        val minL = minOf(lowerBounds[i] - minU, totalD * 0.4)
        
        r[0][i] = minU
        r[1][i] = maxOf(0.0, minL)
        
        // Ensure constraint satisfaction
        val sum = (0 until L).filter { it >= i }.sumOf { r[0][it] } +
                 (0 until L).filter { it <= i }.sumOf { r[1][it] }
        
        if (abs(sum - totalD) > 1e-6) {
            // Adjust to satisfy constraint
            val adjustment = (totalD - sum) / 2.0
            r[0][i] += adjustment
            r[1][i] += adjustment
        }
    }
    
    return r
}

/**
 * Compute variance of MMPP(2) at given time scale
 */
private fun computeMmppVariance(mmpp: MatrixCell, t: Double): Double {
    // Simplified variance computation
    val D0 = mmpp[0]
    val D1 = mmpp[1]
    
    val lambda1 = D1[0, 0]
    val lambda2 = D1[1, 1]
    val r12 = D0[0, 1]
    val r21 = D0[1, 0]
    
    val meanRate = (lambda1 * r21 + lambda2 * r12) / (r12 + r21)
    val variance = meanRate * t * (1.0 + (lambda1 - lambda2).pow(2) / ((r12 + r21).pow(2)))
    
    return variance
}

/**
 * Wrapper for existing M3PP fitting function (assumed to exist)
 */
private fun m3pp22_fitc_approx_cov_multiclass(
    mmpp: MatrixCell,
    classRates: DoubleArray,
    covariance: Double,
    t: Double
): MatrixCell {
    // This would call the existing implementation
    // For now, create a simplified M3PP structure
    val m3pp = MatrixCell(4) // D0, D1, D1_class1, D1_class2
    
    // Copy MMPP structure
    m3pp[0] = mmpp[0]
    m3pp[1] = mmpp[1]
    
    // Split classes proportionally
    val totalRate = classRates.sum()
    if (totalRate > 0) {
        val prop1 = classRates[0] / totalRate
        val prop2 = classRates[1] / totalRate
        
        m3pp[2] = mmpp[1].scale(prop1)
        m3pp[3] = mmpp[1].scale(prop2)
    } else {
        m3pp[2] = Matrix(2, 2)
        m3pp[3] = Matrix(2, 2)
    }
    
    return m3pp
}

/**
 * Extension function to scale a Matrix
 */
private fun Matrix.scale(factor: Double): Matrix {
    val scaled = Matrix(this.getNumRows(), this.getNumCols())
    for (i in 0 until this.getNumRows()) {
        for (j in 0 until this.getNumCols()) {
            scaled[i, j] = this[i, j] * factor
        }
    }
    return scaled
}
/**
 * M3Pp22 Interleave Fit Count algorithms
 */
@Suppress("unused")
class M3pp22InterleaveFitCountAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}